### PR TITLE
Delegate error handling to command framework instead of graph formatters

### LINKED ISSUE #16393: [bug] Consistent command results and error messages accross install/create and graph
### What is your suggestion?

When a dependency is `missing` or `invalid`, CLI commands `install`, `create`, etc, will output pertinent error messages and the exit code of the application will be `!=0`, meaning the command failed.

Making some tests, we have found an inconsistency all along `conan graph` subcommands.
`conan graph info` and `conan graph build-order` should output the same errors as `install` does and should fail in the same way. This does not imply when an error is found not generating the graph output. Commands should generate same graph output but they should fail.

## Example

```sh
 conan install --requires "unleash-client-cpp/1.3.0"
```

<details>

```

======== Input profiles ========
Profile host:
[settings]
arch=armv8
build_type=Release
compiler=apple-clang
compiler.cppstd=gnu17
compiler.libcxx=libc++
compiler.version=15
os=Macos

Profile build:
[settings]
arch=armv8
build_type=Release
compiler=apple-clang
compiler.cppstd=gnu17
compiler.libcxx=libc++
compiler.version=15
os=Macos


======== Computing dependency graph ========
Graph root
    cli
Requirements
    cpr/1.10.5#a227f2a4f0deed3ee1f8b8237477129e - Cache
    libcurl/8.6.0#752cd9c3df0e9487b72301862f15a965 - Cache
    nlohmann_json/3.11.3#45828be26eb619a2e04ca517bb7b828d - Cache
    unleash-client-cpp/1.3.0#a4354a616cd5bdb81354825229954522 - Cache
    zlib/1.3.1#f52e03ae3d251dec704634230cd806a2 - Cache
Build requirements
    autoconf/2.71#f9307992909d7fb3df459340f1932809 - Cache
    automake/1.16.5#058bda3e21c36c9aa8425daf3c1faf50 - Cache
    gnu-config/cci.20210814#dc430d754f465e8c74463019672fb97b - Cache
    libtool/2.4.7#08316dad5c72c541ed21e039e4cf217b - Cache
    m4/1.4.19#b38ced39a01e31fef5435bc634461fd2 - Cache
    meson/1.2.2#04bdfb85d665c82b08a3510aee3ffd19 - Cache
    ninja/1.11.1#77587f8c8318662ac8e5a7867eb4be21 - Cache
    pkgconf/2.1.0#27f44583701117b571307cf5b5fe5605 - Cache
Resolved version ranges
    libcurl/[>=7.78.0 <9]: libcurl/8.6.0
    zlib/[>=1.2.11 <2]: zlib/1.3.1

======== Computing necessary packages ========
cpr/1.10.5: Compatible package ID 01269d344dd5584f4884dfbce8b9b77ae08480e6 equal to the default package ID: Skipping it.
cpr/1.10.5: Checking 5 compatible configurations
cpr/1.10.5: Compatible configurations not found in cache, checking servers
cpr/1.10.5: 'dfee4b06bfe93e4207413fffd653ec0bb2ef0d5c': compiler.cppstd=17
cpr/1.10.5: '1b581912b6a996992095ec289d8a10caf974e2df': compiler.cppstd=20
cpr/1.10.5: 'b59d6a6f26a2d7f87370035527ef219122bb9093': compiler.cppstd=gnu20
cpr/1.10.5: '7b4224207f3ce01b81f5d00add4296667a22e2ba': compiler.cppstd=23
cpr/1.10.5: '839343c67d23729138e2ac0c4c4386e281cd1ffc': compiler.cppstd=gnu23
unleash-client-cpp/1.3.0: Compatible package ID e73661f3afc6efaa3f8ba17b9263163c204f8c52 equal to the default package ID: Skipping it.
unleash-client-cpp/1.3.0: Checking 5 compatible configurations
unleash-client-cpp/1.3.0: Compatible configurations not found in cache, checking servers
unleash-client-cpp/1.3.0: '6332c366c6c529367a2b716abd77f657b96d3c47': compiler.cppstd=17
unleash-client-cpp/1.3.0: '9194788030461f5d772c319ed4b3d745e75e2d47': compiler.cppstd=20
unleash-client-cpp/1.3.0: 'ee3dc95e09fa0d284e807bdf988a10a5cc6e59c9': compiler.cppstd=gnu20
unleash-client-cpp/1.3.0: '4ff0d71ebb4d6a0900ca4e36b2b780168848c458': compiler.cppstd=23
unleash-client-cpp/1.3.0: 'dafb9907f23cc61560a25535b18340c225627896': compiler.cppstd=gnu23
Requirements
    cpr/1.10.5#a227f2a4f0deed3ee1f8b8237477129e:01269d344dd5584f4884dfbce8b9b77ae08480e6 - Missing
    libcurl/8.6.0#752cd9c3df0e9487b72301862f15a965:267b943cf33985f9448f72c97e42fcdb5094fcf5 - Missing
    unleash-client-cpp/1.3.0#a4354a616cd5bdb81354825229954522:e73661f3afc6efaa3f8ba17b9263163c204f8c52 - Missing
    zlib/1.3.1#f52e03ae3d251dec704634230cd806a2:2ee39e692ca4177b4b689b15bc1f2cfdf8f83706 - Missing
Build requirements
Skipped binaries
    nlohmann_json/3.11.3, autoconf/2.71, automake/1.16.5, gnu-config/cci.20210814, libtool/2.4.7, m4/1.4.19, meson/1.2.2, ninja/1.11.1, pkgconf/2.1.0
ERROR: Missing binary: cpr/1.10.5:01269d344dd5584f4884dfbce8b9b77ae08480e6
ERROR: Missing binary: libcurl/8.6.0:267b943cf33985f9448f72c97e42fcdb5094fcf5
ERROR: Missing binary: unleash-client-cpp/1.3.0:e73661f3afc6efaa3f8ba17b9263163c204f8c52
ERROR: Missing binary: zlib/1.3.1:2ee39e692ca4177b4b689b15bc1f2cfdf8f83706

zlib/1.3.1: WARN: Can't find a 'zlib/1.3.1' package binary '2ee39e692ca4177b4b689b15bc1f2cfdf8f83706' for the configuration:
[settings]
arch=armv8
build_type=Release
compiler=apple-clang
compiler.version=15
os=Macos
[options]
fPIC=True
shared=False

ERROR: Missing prebuilt package for 'cpr/1.10.5', 'libcurl/8.6.0', 'unleash-client-cpp/1.3.0', 'zlib/1.3.1'. You can try:
    - List all available packages using 'conan list zlib/1.3.1:* -r=remote'
    - Explain missing binaries: replace 'conan install ...' with 'conan graph explain ...'
    - Try to build locally from sources using the '--build=cpr/1.10.5 --build=libcurl/8.6.0 --build=unleash-client-cpp/1.3.0 --build=zlib/1.3.1' argument

More Info at 'https://docs.conan.io/2/knowledge/faq.html#error-missing-prebuilt-package'

```
</details>

```sh
conan graph info --requires "unleash-client-cpp/1.3.0" > /dev/null
```

<details>

```
======== Computing dependency graph ========
Graph root
    cli
Requirements
    cpr/1.10.5#a227f2a4f0deed3ee1f8b8237477129e - Cache
    libcurl/8.6.0#752cd9c3df0e9487b72301862f15a965 - Cache
    nlohmann_json/3.11.3#45828be26eb619a2e04ca517bb7b828d - Cache
    unleash-client-cpp/1.3.0#a4354a616cd5bdb81354825229954522 - Cache
    zlib/1.3.1#f52e03ae3d251dec704634230cd806a2 - Cache
Build requirements
    autoconf/2.71#f9307992909d7fb3df459340f1932809 - Cache
    automake/1.16.5#058bda3e21c36c9aa8425daf3c1faf50 - Cache
    gnu-config/cci.20210814#dc430d754f465e8c74463019672fb97b - Cache
    libtool/2.4.7#08316dad5c72c541ed21e039e4cf217b - Cache
    m4/1.4.19#b38ced39a01e31fef5435bc634461fd2 - Cache
    meson/1.2.2#04bdfb85d665c82b08a3510aee3ffd19 - Cache
    ninja/1.11.1#77587f8c8318662ac8e5a7867eb4be21 - Cache
    pkgconf/2.1.0#27f44583701117b571307cf5b5fe5605 - Cache
Resolved version ranges
    libcurl/[>=7.78.0 <9]: libcurl/8.6.0
    zlib/[>=1.2.11 <2]: zlib/1.3.1

======== Computing necessary packages ========
cpr/1.10.5: Compatible package ID 01269d344dd5584f4884dfbce8b9b77ae08480e6 equal to the default package ID: Skipping it.
cpr/1.10.5: Checking 5 compatible configurations
cpr/1.10.5: Compatible configurations not found in cache, checking servers
cpr/1.10.5: 'dfee4b06bfe93e4207413fffd653ec0bb2ef0d5c': compiler.cppstd=17
cpr/1.10.5: '1b581912b6a996992095ec289d8a10caf974e2df': compiler.cppstd=20
cpr/1.10.5: 'b59d6a6f26a2d7f87370035527ef219122bb9093': compiler.cppstd=gnu20
cpr/1.10.5: '7b4224207f3ce01b81f5d00add4296667a22e2ba': compiler.cppstd=23
cpr/1.10.5: '839343c67d23729138e2ac0c4c4386e281cd1ffc': compiler.cppstd=gnu23
unleash-client-cpp/1.3.0: Compatible package ID e73661f3afc6efaa3f8ba17b9263163c204f8c52 equal to the default package ID: Skipping it.
unleash-client-cpp/1.3.0: Checking 5 compatible configurations
unleash-client-cpp/1.3.0: Compatible configurations not found in cache, checking servers
unleash-client-cpp/1.3.0: '6332c366c6c529367a2b716abd77f657b96d3c47': compiler.cppstd=17
unleash-client-cpp/1.3.0: '9194788030461f5d772c319ed4b3d745e75e2d47': compiler.cppstd=20
unleash-client-cpp/1.3.0: 'ee3dc95e09fa0d284e807bdf988a10a5cc6e59c9': compiler.cppstd=gnu20
unleash-client-cpp/1.3.0: '4ff0d71ebb4d6a0900ca4e36b2b780168848c458': compiler.cppstd=23
unleash-client-cpp/1.3.0: 'dafb9907f23cc61560a25535b18340c225627896': compiler.cppstd=gnu23
Requirements
    cpr/1.10.5#a227f2a4f0deed3ee1f8b8237477129e:01269d344dd5584f4884dfbce8b9b77ae08480e6 - Missing
    libcurl/8.6.0#752cd9c3df0e9487b72301862f15a965:267b943cf33985f9448f72c97e42fcdb5094fcf5 - Missing
    unleash-client-cpp/1.3.0#a4354a616cd5bdb81354825229954522:e73661f3afc6efaa3f8ba17b9263163c204f8c52 - Missing
    zlib/1.3.1#f52e03ae3d251dec704634230cd806a2:2ee39e692ca4177b4b689b15bc1f2cfdf8f83706 - Missing
Build requirements
Skipped binaries
    nlohmann_json/3.11.3, autoconf/2.71, automake/1.16.5, gnu-config/cci.20210814, libtool/2.4.7, m4/1.4.19, meson/1.2.2, ninja/1.11.1, pkgconf/2.1.0

-------- Basic graph information --------
```
</details>

```sh
conan graph build-order --order-by=recipe --requires "unleash-client-cpp/1.3.0" > /dev/null
```

<details>

```

======== Computing dependency graph ========
Graph root
    cli
Requirements
    cpr/1.10.5#a227f2a4f0deed3ee1f8b8237477129e - Cache
    libcurl/8.8.0#9170903d66d6f6e6795fbb566a403129 - Cache
    nlohmann_json/3.11.3#45828be26eb619a2e04ca517bb7b828d - Cache
    unleash-client-cpp/1.3.0#a4354a616cd5bdb81354825229954522 - Cache
    zlib/1.3.1#f52e03ae3d251dec704634230cd806a2 - Cache
Build requirements
    autoconf/2.71#f9307992909d7fb3df459340f1932809 - Cache
    automake/1.16.5#058bda3e21c36c9aa8425daf3c1faf50 - Cache
    gnu-config/cci.20210814#dc430d754f465e8c74463019672fb97b - Cache
    libtool/2.4.7#08316dad5c72c541ed21e039e4cf217b - Cache
    m4/1.4.19#b38ced39a01e31fef5435bc634461fd2 - Cache
    meson/1.2.2#04bdfb85d665c82b08a3510aee3ffd19 - Cache
    ninja/1.11.1#77587f8c8318662ac8e5a7867eb4be21 - Cache
    pkgconf/2.1.0#27f44583701117b571307cf5b5fe5605 - Cache
Resolved version ranges
    libcurl/[>=7.78.0 <9]: libcurl/8.8.0
    zlib/[>=1.2.11 <2]: zlib/1.3.1

======== Computing necessary packages ========
cpr/1.10.5: Compatible package ID 6da29aec5517b05e82cc36e7f0cea25b756592a1 equal to the default package ID: Skipping it.
cpr/1.10.5: Checking 5 compatible configurations
cpr/1.10.5: Compatible configurations not found in cache, checking servers
cpr/1.10.5: '7c482b127e0f60686b2846e9000db88c8e11666f': compiler.cppstd=17
cpr/1.10.5: '8190a0a48706aaa6e73601f80ccad90dfb634831': compiler.cppstd=20
cpr/1.10.5: '033edebf48b9a193f4775cf79a226f42000d28f2': compiler.cppstd=gnu20
cpr/1.10.5: '4e3c016827f1e0890c20a8270f1157bc59399749': compiler.cppstd=23
cpr/1.10.5: 'a938cbada2ed4cf609e1017660a982a2d845dfe8': compiler.cppstd=gnu23
unleash-client-cpp/1.3.0: Compatible package ID 31826f3ddae6dc65d0a7c1f89aee57b356b02c35 equal to the default package ID: Skipping it.
unleash-client-cpp/1.3.0: Checking 5 compatible configurations
unleash-client-cpp/1.3.0: Compatible configurations not found in cache, checking servers
unleash-client-cpp/1.3.0: '075884ad7d4ee0b37c12ff00532b073d1d95a6ed': compiler.cppstd=17
unleash-client-cpp/1.3.0: 'd1b62f23ae496268defae946797d055b11f8ad07': compiler.cppstd=20
unleash-client-cpp/1.3.0: '1d89d0c5293777c413984d998ab350e8c256ff3f': compiler.cppstd=gnu20
unleash-client-cpp/1.3.0: '8dbe067233297ceac6a97a9cfdc6f0511d43e42d': compiler.cppstd=23
unleash-client-cpp/1.3.0: '0d0154f2b3f41e640700de62e84302ebd7f3ad54': compiler.cppstd=gnu23
Requirements
    cpr/1.10.5#a227f2a4f0deed3ee1f8b8237477129e:6da29aec5517b05e82cc36e7f0cea25b756592a1 - Missing
    libcurl/8.8.0#9170903d66d6f6e6795fbb566a403129:da91ff7ebbce6962ceee923568a6793fa42e7ebe - Missing
    unleash-client-cpp/1.3.0#a4354a616cd5bdb81354825229954522:31826f3ddae6dc65d0a7c1f89aee57b356b02c35 - Missing
    zlib/1.3.1#f52e03ae3d251dec704634230cd806a2:2ee39e692ca4177b4b689b15bc1f2cfdf8f83706 - Missing
Build requirements
Skipped binaries
    nlohmann_json/3.11.3, autoconf/2.71, automake/1.16.5, gnu-config/cci.20210814, libtool/2.4.7, m4/1.4.19, meson/1.2.2, ninja/1.11.1, pkgconf/2.1.0

======== Computing the build order ========
zlib/1.3.1#f52e03ae3d251dec704634230cd806a2:2ee39e692ca4177b4b689b15bc1f2cfdf8f83706 - Missing
libcurl/8.8.0#9170903d66d6f6e6795fbb566a403129:da91ff7ebbce6962ceee923568a6793fa42e7ebe - Missing
cpr/1.10.5#a227f2a4f0deed3ee1f8b8237477129e:6da29aec5517b05e82cc36e7f0cea25b756592a1 - Missing
unleash-client-cpp/1.3.0#a4354a616cd5bdb81354825229954522:31826f3ddae6dc65d0a7c1f89aee57b356b02c35 - Missing
```
</details>

As it can be easily seen, the graph info and build-order commands does not output the expected errors.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

### PR DESCRIPTION
**Pull request by perseoGI**
Changelog: Bugfix: Make conan graph <subcommand> a real "install" dry-run. Return same errors and give same messages
Docs: omit

Close #16393 

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop2/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one.

### CODE REVIEW

File: conan/cli/commands/graph.py [hunk 1]
```diff
@@ -156,7 +159,9 @@ def graph_build_order_merge(conan_api, parser, subparser, *args):
     install_order_serialized = result.install_build_order()
     if getattr(result, "legacy"):
         install_order_serialized = install_order_serialized["order"]
-    return install_order_serialized
+
+    return {"build_order": install_order_serialized,
+            "conan_error": result.get_errors()}
```
Comment (memsharded) [inline conan/cli/commands/graph.py:167]: Is this even possible? How can a build-order-merge have errors?

**Review by memsharded — APPROVED** [review]
Looking good 👍

File: conan/cli/command.py [hunk 2]
```diff
@@ -98,6 +98,14 @@ def _format(self, parser, info, *args):
...
         formatter(info)
 
+    def _raise_errors(self, info):
+        if info and "conan_error" in info and isinstance(info["conan_error"], Exception):
+            raise info["conan_error"]
+        if info and "conan_warning" in info and isinstance(info["conan_warning"], Exception):
+            try:
+                raise info["conan_warning"]
+            except ConanException as e:
+                ConanOutput().warning(e)
```
Comment (memsharded) [inline conan/cli/command.py]: This reads a bit weird, why not printing directly the warning?

File: conan/cli/command.py [hunk 3]
```diff
@@ -98,6 +98,15 @@ def _format(self, parser, info, *args):
 
         formatter(info)
 
+    def _raise_errors(self, info):
```
Comment (memsharded) [inline conan/cli/command.py]: Name is not very correct, if it also prints warnings and it doesn't raise

File: conan/cli/command.py [hunk 4]
```diff
@@ -98,6 +98,15 @@ def _format(self, parser, info, *args):
...
+        if info and type(info) == dict:
+            if "conan_error" in info and isinstance(info["conan_error"], Exception):
+                raise info["conan_error"]
+            if "conan_warning" in info and isinstance(info["conan_warning"], Exception):
+                try:
+                    raise info["conan_warning"]
+                except ConanException as e:
+                    ConanOutput().warning(e)
```
Comment (memsharded) [inline conan/cli/command.py]: This expression is difficult to understand, I'd expect to just print the warnings.

File: conan/cli/command.py [hunk 5]
```diff
@@ -98,6 +98,15 @@ def _format(self, parser, info, *args):
 
         formatter(info)
 
+    def _raise_errors(self, info):
+        if info and type(info) == dict:
+            if "conan_error" in info and isinstance(info["conan_error"], Exception):
```
Comment (memsharded) [inline conan/cli/command.py]: The ``isinstance()`` might be just hiding user errors. If some code defines a ``conan_error``, but it is just a string, this will do nothing.

File: conan/cli/command.py [hunk 6]
```diff
@@ -98,6 +98,15 @@ def _format(self, parser, info, *args):
 
         formatter(info)
 
+    def _raise_errors(self, info):
+        if info and type(info) == dict:
```
Comment (memsharded) [inline conan/cli/command.py]:
```suggestion
        if info and isinstance(info, dict):
```

File: conan/cli/commands/graph.py [hunk 7]
```diff
@@ -234,7 +239,10 @@ def graph_info(conan_api, parser, subparser, *args):
     return {"graph": deps_graph,
             "field_filter": args.filter,
             "package_filter": args.package_filter,
-            "conan_api": conan_api}
+            "conan_api": conan_api,
+            "conan_error": deps_graph.error,
+            # Do not compute graph errors if there are dependency errors
+            "conan_warning": InstallGraph(deps_graph).get_errors() if not deps_graph.error else None}
```
Comment (memsharded) [inline conan/cli/commands/graph.py:248]: This ``conan_warning: ...get_error()`` is a bit weird, it should probably be converted to string at this point

File: conans/client/graph/install_graph.py [hunk 8]
```diff
@@ -454,47 +454,69 @@ def install_build_order(self):
...
-                if package.binary == BINARY_MISSING:
-                    missing.append(package)
-                elif package.binary == BINARY_INVALID:
-                    invalid.append(package)
+        for _, install_node in self._nodes.items():
+            if self._order == "recipe":
+                for package in install_node.packages.values():
+                    analyze_package(package)
+            elif self._order == "configuration":
+                analyze_package(install_node)
```
Comment (memsharded) [inline conans/client/graph/install_graph.py:469]: This is a different bug fix, isn't it? In these cases it makes sense to move it to its own PR, specially when so close to release date, so it can be in the release irrespective if the other feature is delayed.

File: conans/client/graph/install_graph.py [hunk 9]
```diff
@@ -454,47 +454,69 @@ def install_build_order(self):
                   "order": [[n.serialize() for n in level] for level in install_order]}
         return result
 
-    def raise_errors(self):
+    def get_missing_invalid_packages(self):
```
Comment (memsharded) [inline conans/client/graph/install_graph.py]: Private to the class?

File: conans/client/graph/install_graph.py [hunk 10]
```diff
@@ -454,47 +454,69 @@ def install_build_order(self):
...
         if missing:
-            self._raise_missing(missing)
+            self._raise_missing(missing, basic_output)
+
+    def get_errors(self):
+        try:
+            self.raise_errors(basic_output=True)
+        except ConanException as e:
+            return str(e)
+        return None
```
Comment (memsharded) [inline conans/client/graph/install_graph.py]: It looks like that instead of modifying the ``_raise_missing`` and doing this, it would be better to craft its own simpler message here.

**Review by memsharded — COMMENTED** [review]
Looking better, getting closer! :)

File: conans/client/graph/install_graph.py (hunk 10 above)
Comment (perseoGI) [reply conans/client/graph/install_graph.py]: But `raise_errors` could potentially raise invalid binaries errors also.

Comment (memsharded) [reply conans/client/graph/install_graph.py]: Yes, sure, it can iterate all packages and add an error message line for each invalid and missing.

File: conans/client/graph/install_graph.py [hunk 11]
```diff
@@ -454,27 +454,62 @@ def install_build_order(self):
...
         return result
 
-    def raise_errors(self):
+    def _get_missing_invalid_packages(self):
+        def analyze_package(package):
+            if package.binary == BINARY_MISSING:
+                missing.append(package)
+            elif package.binary == BINARY_INVALID:
+                invalid.append(package)
         missing, invalid = [], []
```
Comment (memsharded) [inline conans/client/graph/install_graph.py]:
```suggestion
        missing, invalid = [], []
        def analyze_package(package):
            if package.binary == BINARY_MISSING:
                missing.append(package)
            elif package.binary == BINARY_INVALID:
                invalid.append(package)
```
Even if the evaluation order of python allows it, I think the variables that are within a closure should at least have been defined before the function is defined, for readability.

**Review by memsharded — APPROVED** [review]
Good, it can be moved forward

File: conan/cli/command.py [hunk 12]
```diff
@@ -98,6 +98,12 @@ def _format(self, parser, info, *args):
 
         formatter(info)
 
+    def _dispatch_errors(self, info):
+        if info and isinstance(info, dict):
+            if "conan_error" in info and info["conan_error"]:
+                raise ConanException(info["conan_error"])
+            if "conan_warning" in info and info["conan_warning"]:
+                ConanOutput().warning(info["conan_warning"])
```
Comment (AbrilRBS) [inline conan/cli/command.py]:
```suggestion
            if info.get("conan_error"):
                raise ConanException(info["conan_error"])
            if info.get("conan_warning"):
                ConanOutput().warning(info["conan_warning"])
```

**Review by AbrilRBS — APPROVED** [review]

**Review by memsharded — APPROVED** [review]

**Review by memsharded — APPROVED** [review]

**Review by memsharded — APPROVED** [review]

**Review by czoido — APPROVED** [review]
